### PR TITLE
Add trace enrichment logic for apm aggregation keys

### PR DESCRIPTION
**Pull request by lahsivjar**
Adds enrichment logic for enriching OTel spans. Only the fields required for [apm-aggregation](https://github.com/elastic/apm-aggregation) are currently added. The enrichment logic is based on https://github.com/elastic/apm-data/tree/main, however, I have tried to keep adhering to the latest semconv for now.

Related to: https://github.com/elastic/opentelemetry-dev/issues/304

### CODE REVIEW

**Comment by ishleenk17** [comment]
@lahsivjar : So we plan to use this library for maintaining any type of compatibility between OTEL and Elastic.
Not just Metrics specific. 
That sounds fine. Shall we remove the metrics specific description in the readme of the repo then ?

File: enrichments/trace/internal/elastic/span.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go:175]: > There is no such things as db transaction type in Elastic APM. The [transaction represents the entry](https://github.com/elastic/apm/blob/main/specs/agents/tracing-transactions.md) to a service - that works out nicely for incoming HTTP calls and e.g. for reading out from a message queue. But e.g. a database call always happens within a transaction - that can't be the entry point into a service - something else always needs to trigger the db call - that's how the Elastic APM data model work.

Thanks for the information here. My first thoughts were that if we instrumented a DB as a service then the transaction type would be `db` but now I realize that is not correct and in such cases, the txn would be w.r.t. the db as a service.

> So... here is my question: do we maybe want to try to solve this problem?

I don't yet grasp the whole issue but I think it would be worth to solve it if we can. I will go over the apm-data issue and create one here if required. We can continue the discussion there.

**Comment by lahsivjar** [comment]
@ishleenk17 I think the decision to use this library for all OTel<>Elastic compatibility was when we named it. We should definitely fix the description in the readme, I will create a followup for that.

File: enrichments/trace/internal/elastic/span.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go]: Fixed via https://github.com/elastic/opentelemetry-lib/pull/46/commits/1d7e19708667760a5ccea344dfd655b8998fac10

File: enrichments/trace/internal/elastic/span.go
Comment (gregkalapos) [reply enrichments/trace/internal/elastic/span.go]: Ahh, that's the next huge topic we need to think about :) That is versioning across different SemConv versions.

`db.name` got renamed to `db.namespace` - can be [seen here](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/deprecated/db.yaml#L69-L74). I see Elasticsearch in v1.26.0 which you linked doesn't have `db.namespace`, but that seems to be an error - [`main` already has it](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/elasticsearch.md) - that's how it should be and `db.namespace` be there in the next version.

This happens as part of the DB stabilization - currently all db attributes are experimental, the next version aims to make most of these stable, but as part of this, there are some changes.

As it seems in SemConv v1.26.0 the renaming of `db.name` to `db.namespace` already happened (except for ES, where it just got removed) and got released in the SemConv spec, but that's just the SemConv spec, I doubt any real implementation implemented that. I'd expect most still send `db.name` - in case of Elasticsearch I'm fairly confident we haven't updated this yet (I in discussion wit the ES clients team to go through the SemConv changes and implement the new version - I'm not aware they implemented any of this yet).

> OTOH, vendor-specific attributes like, [db.elasticsearch.cluster.name](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/attributes-registry/db.md), feel a bit brittle to handle.

Agree, that can explode quickly and the goal of SemConv is to have common attributes across vendors (and only have vendor specific ones where it's really needed), so we should take advantage of this and use the vendor agnostic attribute.

> Currently, the logic to determine service target uses db.name but if such attributes are not expected in the semconv maybe we should expand the definition.

Ultimately I think we should have a better story for version handling. In this case, depending on the version of the schema url that's sent, it should be either `db.name` (old version) or `db.namespace` from version `1.26` and higher.

To unblock the PR: since I think no instrumentation implemented `db.namespace` yet, I think having `db.name` for now is fine. And then I'd handle versioning in an extra issue/PR later. 

Alternatively you could check for both - that'd work and one should be always present.

Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go]: > db.name got renamed to db.namespace - can be [seen here](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/deprecated/db.yaml#L69-L74). I see Elasticsearch in v1.26.0 which you linked doesn't have db.namespace, but that seems to be an error - [main already has it](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/elasticsearch.md) - that's how it should be and db.namespace be there in the next version.

Hmm, I see. I have been working with v1.25.0 of the semconv (the latest released one with v0.104.0) so I didn't even see that db.name has been deprecated. Thanks for pointing me to it.

> To unblock the PR: since I think no instrumentation implemented db.namespace yet, I think having db.name for now is fine. And then I'd handle versioning in an extra issue/PR later.

> Alternatively you could check for both - that'd work and one should be always present.

It would probably be best to handle this in a followup PR after we have the v1.26.0 of the semconv released.

File: enrichments/trace/internal/elastic/span_test.go
Comment (lahsivjar) [inline enrichments/trace/internal/elastic/span_test.go:429]: @gregkalapos @axw What are your thoughts on this? As in this test case, we won't fallback to setting the svc target attributes in cases where span represents a db type over HTTP BUT the db attributes are not enough to infer all required information.

**Review by ishleenk17 — APPROVED** [review]
Code owner approval.

File: enrichments/trace/internal/elastic/span.go
Comment (axw) [inline enrichments/trace/internal/elastic/span.go:71]: Given that this is the only accessible part of the API, I think it may be worth changing the API a bit to be like:

```go
func EnrichSpan(span ptrace.Span) {
	var c spanEnrichmentContext
	c.enrich(span)
}

type spanEnrichmentContext {
	urlFull *url.URL

	peerService     string
	...
}

func (c *spanEnrichmentContext) enrich(span ptrace.Span) {
	...
}
```

File: enrichments/trace/internal/elastic/span.go
Comment (axw) [reply enrichments/trace/internal/elastic/span.go]: If we're going to do this, does it belong here? I think maybe it would belong to the same processor, but adding artificial spans to the trace seems fairly different to what we're doing in this code.

File: enrichments/trace/internal/elastic/span.go
Comment (axw) [inline enrichments/trace/internal/elastic/span.go:171]:
```suggestion
```

handled below right?

File: enrichments/trace/internal/elastic/span.go
Comment (axw) [inline enrichments/trace/internal/elastic/span.go:214]: Looks like there's a slight discrepancy with apm-data here: in apm-data we only the the outcome based on HTTP status code if the span status is unset.

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:114]: This comment is a bit confusing to me: I think what you're saying is that the status code takes precedence over span status for setting the `transaction.result`? But OTOH it's the opposite for `event.outcome`. Maybe clarify that in the comment?

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:238]:
```suggestion
				// peer.service should be ignored if more specific deductions
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:334]:
```suggestion
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:366]:
```suggestion
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:302]:
```suggestion
```

Does this comment belong here? It seems that `peer.service` is honoured, per the enriched attributes.

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:283]:
```suggestion
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:261]:
```suggestion
				// peer.service should be ignored if more specific deductions
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:220]:
```suggestion
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [inline enrichments/trace/internal/elastic/span_test.go:383]:
```suggestion
```

File: enrichments/trace/internal/elastic/span_test.go
Comment (axw) [reply enrichments/trace/internal/elastic/span_test.go:429]: I would be inclined to go with bug-for-bug compatibility here for now, and create an issue and follow up. I think it probably makes sense to have an order of precedence like DB > RPC > HTTP > general, both for setting the target type and target name. OTOH it's kinda awkward that one attribute would be set from DB, and another from RPC/HTTP, so it's not 100% obvious to me.

**Review by axw — COMMENTED** [review]
Looks great overall, just a bunch of nit picks.

File: enrichments/trace/internal/elastic/span_test.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span_test.go]: Copy paste fiasco 😮‍💨

File: enrichments/trace/internal/elastic/span_test.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span_test.go]: Fixed! Thanks for finding that!

File: enrichments/trace/internal/elastic/span.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go]: Fixed!

File: enrichments/trace/internal/elastic/span.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go:272]: It would be different then what we do in apm-data right? Is it okay to create a follow-up for this?

Comment (gregkalapos) [reply enrichments/trace/internal/elastic/span.go:272]: > We should ideally decide based on whether the span's parent is remote.

Hmm - very likely I miss something here, because I'm surprised by this. I thought this method would make the decision if this is a local transaction - so if this is the very first span (root span) in the given service. If there is no upstream service, then `isRemote` will still be `false`, but we'd still need to map that root span into a transaction.

> It would be different then what we do in apm-data right? I

Yeah, that's also [how I read it.](https://github.com/elastic/apm-data/blob/main/input/otlp/traces.go#L198)

File: enrichments/trace/internal/elastic/span_test.go
Comment (gregkalapos) [reply enrichments/trace/internal/elastic/span_test.go]: ➕ on the suggestion. I think the core issue is that some of the fields we use here are not required (e.g. `db.name`), so this is not only a theoretical problem - this can happen while the data is 100% compatible with SemConv.

If there is an issue, I'm happy to discuss how we could improve it.

**Review by gregkalapos — APPROVED** [review]
Looks good to me for the points I raised. Nice PR! 👍

File: enrichments/trace/internal/elastic/span.go
Comment (lahsivjar) [reply enrichments/trace/internal/elastic/span.go:272]: I have opened a followup PR for this: https://github.com/elastic/opentelemetry-lib/pull/48/commits/fe1da3a8821d5138cf65fdc6deec0ea24f3ff8d3

We can continue the discussion there.

Comment (axw) [reply enrichments/trace/internal/elastic/span.go:272]: > If there is no upstream service, then isRemote will still be false, but we'd still need to map that root span into a transaction.

Sorry, clarifying: I meant instead of checking the span kind, rather than whether it's the root span.

**Review by axw — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
